designcompose/src/main/java/com/android/designcompose/VariableManager.kt

```diff
@@ -0,0 +1,292 @@
+            // Use the material theme override if one exists
+            val materialColor = MaterialThemeValues.getColor(name, c.name, variableState)
+            materialColor?.let {
+                return it
+            }
+
+            val modeName = variableState.varModeValues?.get(c.name)
+            val modeId =
+                modeName?.let { mName -> c.mode_name_hash[mName] }
+                    ?: c.mode_id_hash[c.default_mode_id]
+                        ?.id // Fallback to default mode in collection
+            modeId?.let { mId ->
+                val value = values_by_mode.values_by_mode[mId]
+                value?.let { vv ->
+                    when (vv) {
+                        is VariableValue.Color -> return vv.value.toColor()
+                        is VariableValue.Alias ->
+                            return resolveVariable(vv.value.id, variableState)
+                                ?.getColor(variableMap, variableState)
+                        else -> return null
+                    }
+                }
+            }
+        }
+        return null
+    }
+
+    // Return this variable's number given the current variable state.
+    private fun Variable.getNumber(
+        variableMap: VariableMap,
+        variableState: VariableState,
+    ): Float? {
+        val collection = variableMap.collections[variable_collection_id]
+        collection?.let { c ->
+            variableState.varModeValues?.let { modes ->
+                val modeName = modes[c.name]
+                val modeId =
+                    modeName?.let { mName -> c.mode_name_hash[mName] }
+                        ?: c.mode_id_hash[
+                                c.default_mode_id] // Fallback to default mode in collection
+                modeId?.let { mId ->
+                    val value = values_by_mode.values_by_mode[mId]
+                    value?.let { vv ->
+                        when (vv) {
+                            is VariableValue.Number -> return vv.value
+                            is VariableValue.Alias ->
+                                return resolveVariable(vv.value.id, variableState)
+                                    ?.getNumber(variableMap, variableState)
+                            else -> return null
+                        }
+                    }
+                }
+            }
+        }
+        return null
+    }
+}
+
+// Return the value out of a NumOrVar enum.
+internal fun NumOrVar.getValue(variableState: VariableState): Float {
+    return when (this) {
+        is NumOrVar.Num -> value
+        is NumOrVar.Var -> VariableManager.getNumber(value, variableState) ?: 0F
+        else -> 0F
+    }
+}
+
+// Return the value of a ColorOrVar enum
+internal fun ColorOrVar.getValue(variableState: VariableState): Color? {
+    return when (this) {
+        is ColorOrVar.Color -> value.toColor()
+        is ColorOrVar.Var -> VariableManager.getColor(value, variableState)
+        else -> null
+    }
+}
```

Comment:
Is there a reason why this code cannot be rewritten with generics? I notice that the color version checks the MaterialThemeValues, but MaterialThemeValues could also support generic types, even if the Google Material Theme only supports colors; it could just return null.  Since we'll also need to implement String and Boolean this seems worth doing, otherwise we'll have four copies of the same functions.

Reply:
I can make a common function to share some code between the getColor() and getNumber() functions, but I don't otherwise see much benefit in trying to use generics. There are only a few types that variables support, and while using generics might save a small bit of code duplication, I don't really think it's worth sacrificing the code readability and simplicity here.